### ISSUE TITLE
Crud Operations Failing for an Map with enum as key

### ISSUE THREAD

**Issue by ujjwal116** [spring-cloud/spring-cloud-gcp#1689]
We have an embedded entity which uses MAP, The key for Map is  Java Enum. While reading(findById) and writing(save) this entity we are receiving a converison exception.

From Spring docs it suggest Enum is supported as a datatype , so can you help understand what's going wrong here.

https://cloud.spring.io/spring-cloud-static/spring-cloud-gcp/1.1.0.RC2/multi/multi__spring_data_cloud_datastore.html

#### Environment details
1. Java version: 8
2. spring-cloud-datastore: 1.1.0.RELEASE

#### Code example

  ``` 
 @Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity(name = "organization_preferences")
public class OrganizationPreferencesDto {
    @Id
    @Field(name = "organization_id")
    private String organizationId;

    @Field(name = "events")
    private Map<String, EventDto> eventNameToEventMapping;

}

``` 
  ```
public class EventDto {
    @Field(name = "event_name")
    private String eventName;

    @Field(name = "preferred_channels")
    private List<CommunicationChannels> preferredChannels;

    @Field(name = "preferences")
    private Map<CommunicationChannels, List<String>> channelNameToPreferences;

}
``` 
```
public enum CommunicationChannels {

	EMAIL(ChannelType.EMAIL, "EMAIL"), SMS(ChannelType.SMS, "SMS");
	
	private ChannelType grpcChannelFormat;
	private String channelName;

	private CommunicationChannels(final ChannelType grpcChannelFormat, final String channelName) {
		this.grpcChannelFormat = grpcChannelFormat;
		this.channelName = channelName;
	}

	public String getChannelName() {
		return channelName;
	}


	public ChannelType getGrpcChannelFormat() {
		return grpcChannelFormat;
	}
}

``` 

#### Stack trace
```
2019-05-22 03:01:10.365  WARN 12586 --- [          Gax-8] c.g.cloud.pubsub.v1.MessageDispatcher    : MessageReceiver failed to processes ack ID: K0MLKlgRTgQhIT4wPkVTRFAGFixdRkhRNxkIaFEOT14jPzUgKEURIysdAGNfNBsNaFxcdAEAWRp7LDRxY1kZAQhFe059ZB0Oa1pdcQZZBRx2e2R1aXLN1YuE6bfgagExULPX0dBkLcny58FEZh09WBJLLD5-LTlFQV5AEkwmDERJUw, the message will be nacked.

org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException: Unable to write organization_preferences.events; nested exception is org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException: Unable to write eventDto.preferences; nested exception is org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException: Unable to convert class com.kramphub.postino.organization.preferences.CommunicationChannels to class java.lang.String
	at org.springframework.cloud.gcp.data.datastore.core.convert.DefaultDatastoreEntityConverter.lambda$write$1(DefaultDatastoreEntityConverter.java:149) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.mapping.DatastorePersistentEntityImpl.lambda$doWithColumnBackedProperties$0(DatastorePersistentEntityImpl.java:110) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.data.mapping.model.BasicPersistentEntity.doWithProperties(BasicPersistentEntity.java:355) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.mapping.DatastorePersistentEntityImpl.doWithColumnBackedProperties(DatastorePersistentEntityImpl.java:107) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.DefaultDatastoreEntityConverter.write(DefaultDatastoreEntityConverter.java:133) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.DefaultDatastoreEntityConverter.write(DefaultDatastoreEntityConverter.java:47) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.DatastoreTemplate.convertToEntityForSave(DatastoreTemplate.java:310) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.DatastoreTemplate.save(DatastoreTemplate.java:115) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.repository.support.SimpleDatastoreRepository.save(SimpleDatastoreRepository.java:80) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
	at org.springframework.data.repository.core.support.RepositoryComposition$RepositoryFragments.invoke(RepositoryComposition.java:359) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.data.repository.core.support.RepositoryComposition.invoke(RepositoryComposition.java:200) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$ImplementationMethodExecutionInterceptor.invoke(RepositoryFactorySupport.java:644) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:608) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.lambda$invoke$3(RepositoryFactorySupport.java:595) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:595) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:59) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:93) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.data.repository.core.support.SurroundingTransactionDetectorMethodInterceptor.invoke(SurroundingTransactionDetectorMethodInterceptor.java:61) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at com.sun.proxy.$Proxy66.save(Unknown Source) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:343) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:139) ~[spring-tx-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at com.sun.proxy.$Proxy66.save(Unknown Source) ~[na:na]
	at com.kramphub.postino.organization.preferences.OrganizationPreferencesService.insertByOrganizationId(OrganizationPreferencesService.java:65) ~[classes/:na]
	at com.kramphub.postino.organization.CreateOrganizationPubSubReader.processMessage(CreateOrganizationPubSubReader.java:51) ~[classes/:na]
	at com.kramphub.postino.organization.CreateOrganizationPubSubReader.lambda$0(CreateOrganizationPubSubReader.java:90) ~[classes/:na]
	at com.google.cloud.pubsub.v1.MessageDispatcher$4.run(MessageDispatcher.java:438) ~[google-cloud-pubsub-1.59.0.jar:1.59.0]
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]
Caused by: org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException: Unable to write eventDto.preferences; nested exception is org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException: Unable to convert class com.kramphub.postino.organization.preferences.CommunicationChannels to class java.lang.String
	at org.springframework.cloud.gcp.data.datastore.core.convert.DefaultDatastoreEntityConverter.lambda$write$1(DefaultDatastoreEntityConverter.java:149) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.mapping.DatastorePersistentEntityImpl.lambda$doWithColumnBackedProperties$0(DatastorePersistentEntityImpl.java:110) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.data.mapping.model.BasicPersistentEntity.doWithProperties(BasicPersistentEntity.java:355) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.mapping.DatastorePersistentEntityImpl.doWithColumnBackedProperties(DatastorePersistentEntityImpl.java:107) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.DefaultDatastoreEntityConverter.write(DefaultDatastoreEntityConverter.java:133) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.DefaultDatastoreEntityConverter.write(DefaultDatastoreEntityConverter.java:47) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.lambda$convertOnWriteSingleEmbedded$5(TwoStepsConversions.java:303) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.applyEntityValueBuilder(TwoStepsConversions.java:282) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnWriteSingleEmbedded(TwoStepsConversions.java:302) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.lambda$convertOnWrite$3(TwoStepsConversions.java:255) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnWrite(TwoStepsConversions.java:275) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.lambda$convertOnWriteSingleEmbeddedMap$4(TwoStepsConversions.java:294) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.applyEntityValueBuilder(TwoStepsConversions.java:282) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnWriteSingleEmbeddedMap(TwoStepsConversions.java:288) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.lambda$convertOnWrite$2(TwoStepsConversions.java:251) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnWrite(TwoStepsConversions.java:275) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnWrite(TwoStepsConversions.java:238) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.DefaultDatastoreEntityConverter.lambda$write$1(DefaultDatastoreEntityConverter.java:137) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	... 49 common frames omitted
Caused by: org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException: Unable to convert class com.kramphub.postino.organization.preferences.CommunicationChannels to class java.lang.String
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnReadSingle(TwoStepsConversions.java:231) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.lambda$convertOnWriteSingleEmbeddedMap$4(TwoStepsConversions.java:291) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.applyEntityValueBuilder(TwoStepsConversions.java:282) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnWriteSingleEmbeddedMap(TwoStepsConversions.java:288) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.lambda$convertOnWrite$2(TwoStepsConversions.java:251) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnWrite(TwoStepsConversions.java:275) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnWrite(TwoStepsConversions.java:238) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.DefaultDatastoreEntityConverter.lambda$write$1(DefaultDatastoreEntityConverter.java:137) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	... 66 common frames omitted


```

#### Additional information.	
We have also tried to use custom converters for same but it did not help while writing
```
public static final Converter<CommunicationChannels, String> COMMCHANNEL_STRING_CONVERTER = new Converter<CommunicationChannels, String>() {

		@Override
		public String convert(CommunicationChannels channel) {
			return channel.name();
		}
	};
	public static final Converter<String, CommunicationChannels> STRING_COMMCHANNEL_CONVERTER = new Converter<String, CommunicationChannels>() {

		@Override
		public CommunicationChannels convert(String channel) {
			if (channel.equalsIgnoreCase(CommunicationChannels.EMAIL.toString()))
				return CommunicationChannels.EMAIL;
			else
				return CommunicationChannels.SMS;
		}
	};

@Bean
	public DatastoreCustomConversions datastoreCustomConversions() {
		return new DatastoreCustomConversions(Arrays.asList(CommunicationChannels.COMMCHANNEL_STRING_CONVERTER,
				CommunicationChannels.STRING_COMMCHANNEL_CONVERTER));
	}
```

Thanks!

**Comment by meltsufin** [comment]
@dmitry-s PTAL.

**Comment by dmitry-s** [comment]
Thanks for reporting the issue!
I was able to replicate it.  Will investigate.

**Comment by ayushj158** [comment]
@dmitry-s : I am a colleague of Ujjwal and we looked into this issue together.

One interesting observation we had is
after we add read and write converters 
1. conversion during read operations works fine (successful able to parse string to enumerate) 
2. Write conversion of List<enum> is successful 
3. only write conversion of Map<enum,Object> fails.


Thought it might help with your investigation.

**Comment by ayushj158** [comment]
@dmitry-s  Thanks for the quick fix. In which version is this fix available to us ?

**Comment by dmitry-s** [comment]
It should be available in 1.2.0.BUILD-SNAPSHOT

**Comment by ujjwal116** [comment]
@dmitry-s When this fix will be released?

**Comment by meltsufin** [comment]
@dmitry-s Was this fix applied to `1.1.x` branch?

**Comment by dmitry-s** [comment]
@meltsufin  Yes, it was
@ujjwal116  It will be released with Spring Cloud Greenwich.SR2 relese (see https://github.com/spring-cloud/spring-cloud-release/milestones for schedule)
